### PR TITLE
Made _locate_spatial_element public, renamed to locate_element()

### LINKED ISSUE #425: `contains_element` returns False if element is contained more than once
SpatialData allows an element to be contained more than once with different names. When checking whether it is contained, the result is unexpectedly False although the element is contained.

## Example

```python
from spatialdata import SpatialData
from spatialdata.models import Image2DModel
import numpy as np
img = Image2DModel.parse(np.arange(4).reshape((1, 2, 2)), dims=("c", "y", "x"))
sdata = SpatialData(images={"img1": img, "img2": img})
assert sdata.contains_element(img)
```

## Cause

`_locate_spatial_element` raises a ValueError if the element is not found and another ValueError if the element is found more than once. `contains_element` catches both and returns False, but it should catch the second case separately and return True.

It's maybe worth to consider whether to disallow the same element under different keys (how would that be represented on disk?), because it is a very special edge case that breaks many other code places of spatialdata (e.g. `_write_transformations_to_disk`) and would need to be completely covered by test cases (if it has already any).

Instead of having `contains_element` inspect the AssertionError string from `locate_spatial_element` to distinguish the two ValueErrors, we could either have ValueError subclasses, or a private helper method that returns the counts of found elements (0, 1, 2…) instead of raising exceptions.

### PR DESCRIPTION
**Pull request by LucaMarconato**
Closes #425.

### CODE REVIEW

File: src/spatialdata/_core/spatialdata.py [hunk 1]
```diff
@@ -341,7 +340,7 @@ def _locate_spatial_element(self, element: SpatialElement) -> tuple[str, str]:
                     found_element_type.append(element_type)
                     found_element_name.append(element_name)
         if len(found) == 0:
-            raise ValueError("Element not found in the SpatialData object.")
+            return None
```
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply src/spatialdata/_core/spatialdata.py:343]: Although, now I think with a named tuple (or class) it makes more sense:
```python
located = sdata.locate_element(element)
if located is None:
    do_something_with(located.name)
```

[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:343]: Thank you for your comment, answering in the other conversation.

File: src/spatialdata/_core/spatialdata.py [hunk 2]
```diff
@@ -352,29 +351,6 @@ def _locate_spatial_element(self, element: SpatialElement) -> tuple[str, str]:
         assert len(found_element_type) == 1
         return found_element_name[0], found_element_type[0]
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:352]: I agree that the current approach is not ideal for the two reasons that you mentioned and it's better to change it. Nevertheless, taking into consideration the planned support for nested NGFF files I think we should change the return type to str and return the full path of an element (e.g. 'image/my_image`). In fact the returned type would lose meaning for nested paths and it's more robust to use `spatialdata.models.get_model` instead. In your case I would get the name using `os.path.basename` and get the type in a forward-compatible way using `get_model()`. Would this work for you or do you have comments on this?

**Review by giovp — CHANGES_REQUESTED** [review]
1) this very small method is used only 2 times by 1 private method (!) and 1 public function, for which behaviour I think I actually disagree with because it clashes with the in-memory/on-disk behaviour we removed in #329 . I would at a minum make it a helper private function in utils, but I understand there is a request for it #423 

2) this is essentially a way to perform a lookup in a container based on the value instead of the key. This can be easily done with a dict comprehension on the various elements. 

@aeisenbarth is there something specific that would prevent you from doing something like this `next(key for key, value in dd.items() if value == 'value')` on the various elements?

If this is really needed, I agree with @aeisenbarth that it should not return `Tuple[str,str] | None` but `str | None` with str being the absolute path in spatialdata (e.g. "images/image1" ).

File: src/spatialdata/_core/spatialdata.py [hunk 3]
```diff
@@ -312,24 +312,23 @@ def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> za
...
         Returns
         -------
-        name and type of the element
+        name and type of the element; if the element is not found, None is returned instead
 
         Raises
         ------
         ValueError
-            the element is not found or found multiple times in the SpatialData object
+            the element is found multiple times in the SpatialData object
```
[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/_core/spatialdata.py:331]: and, precisely because there is nothing wrong about storing 2 exact same images in 2 different keys, that such odd behaviors should be left to the users and not being a method provided in the API imho

File: src/spatialdata/_core/spatialdata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply src/spatialdata/_core/spatialdata.py:352]: Yes.
And regarding returning only the full path (`"image/my_image"`), we would then need the ability to do lookups by path, which are currently `sdata.images["my_image"]` or `sdata["my_image"]`. I'm not sure if keeping both the new and old way would be inconsistent and make the code unnecessarily complex.

**Comment by aeisenbarth** [comment]
> is there something specific that would prevent you from doing something like this `next(key for key, value in dd.items() if value == 'value')` on the various elements?

Not really, except that on first try I did exactly that and got a deep comparison of every nested xarray attribute and array chunk, which was a major performance hit. Now knowing SpatialData's private implementation, I can do:
```python
next(name for name, image in sdata.images.items() if id(image) == id(element))
```
which I then would refactor out to a utils function, which is so generic (not specific to my module) that one would wonder whether it better lives in SpatialData itself. Especially since other users would implement the same utils function, possibly doing the same mistakes.

But using that code is fine for me!

**Comment by giovp** [comment]
thanks @aeisenbarth 

> which I then would refactor out to a utils function, which is so generic (not specific to my module) that one would wonder whether it better lives in SpatialData itself. Especially since other users would implement the same utils function, possibly doing the same mistakes.

this is a good enough use case. I would then add the method to both `Elements` but also `SpatialData`, wdyt @LucaMarconato ?

File: src/spatialdata/_core/spatialdata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:352]: Eventually the lookup by path will be supported. Users will still be able to lookup things in a backward compatible way for a while and they will be either able to change the code, or to change the data in a flat storage.

**Comment by LucaMarconato** [comment]
Thanks for the comments, sounds good to me.

@aeisenbarth I would return paths; I suggest to get the name of the path for the lookup until we implement lookup by path. To get the type of the element I recommend to use `get_model()`.

File: src/spatialdata/_core/spatialdata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:331]: I would discourage it because as @aeisenbarth pointed out it could lead to weird behavior downstream. Here I do the check so we ensure that the return path is non-ambiguous.

**Comment by LucaMarconato** [comment]
@giovp ready for re-review (please see also my comment in the remaining unresolved conversation).

File: src/spatialdata/_core/spatialdata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:331]: Also Wouter's comment here: https://github.com/scverse/spatialdata/issues/425#issuecomment-1883408895

[the remark this answers is not in the document]
Comment (giovp) [reply src/spatialdata/_core/spatialdata.py:331]: sorry @LucaMarconato but I think the amount of complexity added on checking the consistency is tbh a lot more than just returning 2 strings here. I think a warning for sure is necessary here, but not a fail, as it should fail earlier.

[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:331]: I was thinking more of returning the first string "trusting" that the object don't contain a copy. I really think that we should not allow at this stage to have duplicate elements. Maybe we could add it later in a separate PR after deep testing, but right now it would the make the library less stable without a real use case.

**Comment by LucaMarconato** [comment]
I have addressed all the comments (changed the return type to an optional list of strings). So ready to merge now.
